<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Anton Enright &amp; Dimitrios Vitsios" />


<title>Introduction to R</title>

<script src="Intro_R_Practical_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="Intro_R_Practical_files/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="Intro_R_Practical_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="Intro_R_Practical_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="Intro_R_Practical_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="Intro_R_Practical_files/navigation-1.1/tabsets.js"></script>
<link href="Intro_R_Practical_files/highlightjs-1.1/textmate.css" rel="stylesheet" />
<script src="Intro_R_Practical_files/highlightjs-1.1/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>



<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Introduction to R</h1>
<h4 class="author"><em>Anton Enright &amp; Dimitrios Vitsios</em></h4>
<h4 class="date"><em>'12 June, 2017'</em></h4>

</div>

<div id="TOC">
<ul>
<li><a href="#introduction-to-r">Introduction to R</a><ul>
<li><a href="#about-r">About R</a></li>
<li><a href="#getting-started">Getting Started</a></li>
<li><a href="#assignment-and-operations">Assignment and Operations</a></li>
<li><a href="#object-types-in-r">Object types in R</a></li>
<li><a href="#questioning-the-numbers">Questioning the numbers</a></li>
<li><a href="#basic-plotting-and-graphing">Basic Plotting and Graphing</a></li>
<li><a href="#exercises">Exercises</a></li>
</ul></li>
</ul>
</div>

<div id="introduction-to-r" class="section level1">
<h1>Introduction to R</h1>
<div id="about-r" class="section level2">
<h2>About R</h2>
<p>R is a free software environment for statistical computing. R is a free tool that you can download and use on a variety of platforms and environments. The <a href="http://www.r-project.org">R project</a> R Webpage is a good place to learn more about R.</p>
<p>In this session we will focus on learning the basics of R so that we will be able to use BioConductor properly in the next session. R contains a vast collection of resources and tools many of which are beyond the scope of this session.</p>
</div>
<div id="getting-started" class="section level2">
<h2>Getting Started</h2>
<p>First load up RStudio, there should be a shortcut on your desktop. You should now see a window like this: <img src="rstudio.png" /></p>
<p>This is the R terminal where we will type commands for both R and Bioconductor.</p>
<p>In this web page, sections like the following contain commands that you can copy and paste into the R terminal. The output you should expect to get from R will be presented like this (below).</p>
<pre class="r"><code>setwd(&quot;/Users/dvitsios/Desktop/EMBO_2017/Courses-and-Practicals/EMBO_Greece_2017/Intro_R&quot;)
print(&quot;Hello Anton!&quot;)</code></pre>
<pre><code>## [1] &quot;Hello Anton!&quot;</code></pre>
<p>To get help in R just type a question mark before the name of the command, for example, for the plot&lt; function:</p>
<pre class="r"><code>?plot</code></pre>
<p>Finally to quit R just type at the prompt:</p>
<pre><code>quit()</code></pre>
</div>
<div id="assignment-and-operations" class="section level2">
<h2>Assignment and Operations</h2>
<p>Simple operations are easy to calculate in R, including multiplication, division, addition, subtraction etc. Here are some examples results below commands:</p>
<pre class="r"><code>2+1

3-5

4*4

16/4

7.15 * sqrt(4)</code></pre>
<pre><code>## [1] 3
## [1] -2
## [1] 16
## [1] 4
## [1] 14.3</code></pre>
<p>To save one of these results, we can assign it to a variable, using the “&lt;-” symbol. We can then use these variables in further operations.</p>
<pre class="r"><code>x &lt;- 2+1

y &lt;- 7.15 * sqrt(4)

x + y</code></pre>
<pre><code>## [1] 17.3</code></pre>
<p>One of R’s benefits is that it can deal with many numbers at once. In order to tell R that we want to treat a series of numbers as one object, we can use the c command to “combine” objects into a list.</p>
<pre class="r"><code>bubba &lt;- c(1,2,3,4,10,59)</code></pre>
<p>To see what numbers are stored in bubba we can just type “bubba”:</p>
<pre class="r"><code>bubba</code></pre>
<pre><code>## [1]  1  2  3  4 10 59</code></pre>
<p>And now, we can perform operations on all of these numbers at once:</p>
<pre class="r"><code>bubba + 1

bubba * 10</code></pre>
<pre><code>## [1]  2  3  4  5 11 60
## [1]  10  20  30  40 100 590</code></pre>
<p>We can also use the many functions available in R:</p>
<pre class="r"><code>sum(bubba)
mean(bubba)
max(bubba)
min(bubba)
sd(bubba)
range(bubba)
length(bubba)</code></pre>
<pre><code>## [1] 79
## [1] 13.16667
## [1] 59
## [1] 1
## [1] 22.67524
## [1]  1 59
## [1] 6</code></pre>
<p>If you don’t know the proper command for a mathematical function you might want to use, you can either use the “Search help” option in the “Help” menu, or the “??” command:</p>
<pre><code>??deviation</code></pre>
<p>If you wish to work with only one of the numbers you can get access to it using the variable and then square brackets indicating the position in the series:</p>
<pre class="r"><code>bubba[1]        # This accesses the first element of bubba

bubba[5]        # The Fifth Element, of bubba</code></pre>
<pre><code>## [1] 1
## [1] 10</code></pre>
<p>Using the <strong>c</strong> command we can also store text (between quotes):</p>
<pre class="r"><code>names &lt;- c(&quot;Tom&quot;, &quot;Anton&quot;, &quot;James&quot;)

names[2]</code></pre>
<pre><code>## [1] &quot;Anton&quot;</code></pre>
<p>You now have a list of numbers and are ready to explore. In the sections that follow we will examine the basic operations in R that will allow you to do some of the analyses required in the next practical.</p>
</div>
<div id="object-types-in-r" class="section level2">
<h2>Object types in R</h2>
<p>These objects are simple lists, we will need later to use more complicated objects to store more complicated data.</p>
<div class="figure">
<img src="images/Rclasses.png" />
</div>
</div>
<div id="questioning-the-numbers" class="section level2">
<h2>Questioning the numbers</h2>
<p>With R, we can “ask” things of the numbers. For instance, we might want to know which of the stored in <em>bubba</em> are larger than 2:</p>
<pre class="r"><code>bubba

bubba &gt; 2</code></pre>
<pre><code>## [1]  1  2  3  4 10 59
## [1] FALSE FALSE  TRUE  TRUE  TRUE  TRUE</code></pre>
<p>We get a list of TRUE / FALSE values (usually called <em>“Logical”</em>). They can be quite useful, as can be seen by the following two examples:</p>
<pre class="r"><code>sum(bubba &gt; 2)

bubba[bubba &gt; 2]</code></pre>
<pre><code>## [1] 4
## [1]  3  4 10 59</code></pre>
<ul>
<li>What do these commands do?</li>
<li>How would you print out the numbers in bubba that are smaller or larger than the <em>mean</em> of bubba?</li>
</ul>
</div>
<div id="basic-plotting-and-graphing" class="section level2">
<h2>Basic Plotting and Graphing</h2>
<p>Once your data is stored in a variable it is relatively easy to plot. To plot the values we just assigned simply type:</p>
<pre class="r"><code>plot(bubba)</code></pre>
<p><img src="Intro_R_Practical_files/figure-html/unnamed-chunk-13-1.png" /><!-- --></p>
<p>You should have a plot like the one above</p>
<p>You could also try other types of plot such as:</p>
<pre class="r"><code>barplot(bubba)</code></pre>
<p><img src="Intro_R_Practical_files/figure-html/unnamed-chunk-14-1.png" /><!-- --></p>
<pre class="r"><code>pie(bubba)</code></pre>
<p><img src="Intro_R_Practical_files/figure-html/unnamed-chunk-14-2.png" /><!-- --></p>
<p>You might also try playing with some example data on flowers (iris) that come with R.</p>
<pre class="r"><code>flowerdata &lt;- iris       # Reads in the included &quot;iris&quot; data and saves it as &quot;flowerdata&quot;</code></pre>
<p>It is quite useful to find out what kind of object you have, and what kind of data it contains:</p>
<pre class="r"><code>class(flowerdata)</code></pre>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<p>A “data.frame” is a table. The flowerdata data.frame contains information about flowers, which we can open in an external editor, or simply view inside R by typing “flowerdata”.</p>
<pre><code>edit(flowerdata)         # Opens an external editor</code></pre>
<p>To quickly find out the size of the table, we can use ask for its dimensions:</p>
<pre class="r"><code>dim(flowerdata)          # returns the dimensions of an object</code></pre>
<pre><code>## [1] 150   5</code></pre>
<p>We now know that flowerdata contains 150 rows and 5 columns. We can use the square brackets to access subsets of the table. In order to access particular cells we need to specify the row number and then the column number separated by a comma. By leaving out either the row or the column number we can gain access to complete columns or rows.</p>
<pre class="r"><code>flowerdata[11,2]         # the element that is in row 11 and column 2</code></pre>
<pre><code>## [1] 3.7</code></pre>
<pre class="r"><code>flowerdata[11,]          # row number 11</code></pre>
<pre><code>##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 11          5.4         3.7          1.5         0.2  setosa</code></pre>
<pre class="r"><code>flowerdata[,2]            # column number 2</code></pre>
<pre><code>##   [1] 3.5 3.0 3.2 3.1 3.6 3.9 3.4 3.4 2.9 3.1 3.7 3.4 3.0 3.0 4.0 4.4 3.9
##  [18] 3.5 3.8 3.8 3.4 3.7 3.6 3.3 3.4 3.0 3.4 3.5 3.4 3.2 3.1 3.4 4.1 4.2
##  [35] 3.1 3.2 3.5 3.6 3.0 3.4 3.5 2.3 3.2 3.5 3.8 3.0 3.8 3.2 3.7 3.3 3.2
##  [52] 3.2 3.1 2.3 2.8 2.8 3.3 2.4 2.9 2.7 2.0 3.0 2.2 2.9 2.9 3.1 3.0 2.7
##  [69] 2.2 2.5 3.2 2.8 2.5 2.8 2.9 3.0 2.8 3.0 2.9 2.6 2.4 2.4 2.7 2.7 3.0
##  [86] 3.4 3.1 2.3 3.0 2.5 2.6 3.0 2.6 2.3 2.7 3.0 2.9 2.9 2.5 2.8 3.3 2.7
## [103] 3.0 2.9 3.0 3.0 2.5 2.9 2.5 3.6 3.2 2.7 3.0 2.5 2.8 3.2 3.0 3.8 2.6
## [120] 2.2 3.2 2.8 2.8 2.7 3.3 3.2 2.8 3.0 2.8 3.0 2.8 3.8 2.8 2.8 2.6 3.0
## [137] 3.4 3.1 3.0 3.1 3.1 3.1 2.7 3.2 3.3 3.0 2.5 3.0 3.4 3.0</code></pre>
<p>We can use different plots to look at the data:</p>
<pre class="r"><code>plot(flowerdata[,2])</code></pre>
<p><img src="Intro_R_Practical_files/figure-html/unnamed-chunk-19-1.png" /><!-- --></p>
<pre class="r"><code>hist(flowerdata[,2])</code></pre>
<p><img src="Intro_R_Practical_files/figure-html/unnamed-chunk-19-2.png" /><!-- --></p>
<pre class="r"><code>boxplot(flowerdata, col=rainbow(5))</code></pre>
<p><img src="Intro_R_Practical_files/figure-html/unnamed-chunk-19-3.png" /><!-- --></p>
<p>In order to save some of these results as a PDF file, we can do the following:</p>
<pre><code>pdf(&quot;flower_images.pdf&quot;)

boxplot(flowerdata, col=rainbow(5))

dev.off()</code></pre>
<p>This will send any image that is produced after the <i>pdf</i> function into a file, closing that file when <em>dev.off()</em> is called.</p>
##Loading Data From a Text File
</H3>
<p>Lets load in some external data from a text file.</p>
<p>The file <a href="&quot;./genomes.txt&quot;">genomes.txt</a> should already be inside the course_material folder on your desktop. We can then load this file into R with the following:</p>
<pre class="r"><code>genomes &lt;- read.table(&quot;genomes.txt&quot;, row.names=1, header=TRUE)</code></pre>
<p>What does this file contain? Let us try out some more commands:</p>
<pre class="r"><code>dim(genomes)</code></pre>
<pre><code>## [1]  9 13</code></pre>
<pre class="r"><code>colnames(genomes)</code></pre>
<pre><code>##  [1] &quot;Genome.length&quot;      &quot;Genes&quot;              &quot;Transcripts&quot;       
##  [4] &quot;Coding&quot;             &quot;Exons&quot;              &quot;Introns&quot;           
##  [7] &quot;Transcripts.length&quot; &quot;Coding.length&quot;      &quot;Exons.length&quot;      
## [10] &quot;Introns.length&quot;     &quot;UTR5.length&quot;        &quot;UTR3.length&quot;       
## [13] &quot;MultiTransGenes&quot;</code></pre>
<pre class="r"><code>rownames(genomes)</code></pre>
<pre><code>## [1] &quot;cat&quot;        &quot;chicken&quot;    &quot;dog&quot;        &quot;drosophila&quot; &quot;human&quot;     
## [6] &quot;mouse&quot;      &quot;opossum&quot;    &quot;worm&quot;       &quot;zebrafish&quot;</code></pre>
<pre class="r"><code>head(genomes)</code></pre>
<pre><code>##            Genome.length Genes Transcripts Coding  Exons Introns
## cat           4055847588 18377       18377  15048 194384  176007
## chicken       1100480441 17858       17858  16736 164238  146380
## dog           2531673953 24045       24045  19305 201708  177663
## drosophila     168736537 15178       15178  14141  60048   44870
## human         3090081323 36953       36953  21359 231277  194324
## mouse         2716965481 31804       31804  23497 217952  186148
##            Transcripts.length Coding.length Exons.length Introns.length
## cat                  25726822      24621516     25726822      827552084
## chicken              29749958      24332411     29749958      376951972
## dog                  33990435      30675983     33990435      587007775
## drosophila           28857046      22370742     28857046       48449314
## human                65116403      33887121     65116403     1060714248
## mouse                59754111      34549583     59754111      873541323
##            UTR5.length UTR3.length MultiTransGenes
## cat                  0           0              NA
## chicken         905240     4349152            3901
## dog             246727     1310786            4524
## drosophila     2339329     3836127            3463
## human          4282528    20893311           11075
## mouse          3750716    17745067            8903</code></pre>
<pre class="r"><code>genomes[&quot;mouse&quot;,]</code></pre>
<pre><code>##       Genome.length Genes Transcripts Coding  Exons Introns
## mouse    2716965481 31804       31804  23497 217952  186148
##       Transcripts.length Coding.length Exons.length Introns.length
## mouse           59754111      34549583     59754111      873541323
##       UTR5.length UTR3.length MultiTransGenes
## mouse     3750716    17745067            8903</code></pre>
<pre class="r"><code>summary(genomes)</code></pre>
<pre><code>##  Genome.length           Genes        Transcripts        Coding     
##  Min.   :1.003e+08   Min.   :15178   Min.   :15178   Min.   :14141  
##  1st Qu.:1.100e+09   1st Qu.:18377   1st Qu.:18377   1st Qu.:16736  
##  Median :2.532e+09   Median :24045   Median :24045   Median :19466  
##  Mean   :2.095e+09   Mean   :24227   Mean   :24227   Mean   :19319  
##  3rd Qu.:3.090e+09   3rd Qu.:27928   3rd Qu.:27928   3rd Qu.:21359  
##  Max.   :4.056e+09   Max.   :36953   Max.   :36953   Max.   :24147  
##                                                                     
##      Exons           Introns       Transcripts.length Coding.length     
##  Min.   : 60048   Min.   : 44870   Min.   :25726822   Min.   :22370742  
##  1st Qu.:164238   1st Qu.:146380   1st Qu.:29018327   1st Qu.:24621516  
##  Median :194384   Median :176007   Median :31700945   Median :30462695  
##  Mean   :178314   Mean   :154086   Mean   :38555971   Mean   :28923326  
##  3rd Qu.:217952   3rd Qu.:186148   3rd Qu.:43089692   3rd Qu.:33887121  
##  Max.   :231277   Max.   :195807   Max.   :65116403   Max.   :34549583  
##                                                                         
##   Exons.length      Introns.length       UTR5.length     
##  Min.   :25726822   Min.   :3.213e+07   Min.   :      0  
##  1st Qu.:29018327   1st Qu.:3.770e+08   1st Qu.: 246727  
##  Median :31700945   Median :5.870e+08   Median : 905240  
##  Mean   :38555971   Mean   :5.824e+08   Mean   :1569940  
##  3rd Qu.:43089692   3rd Qu.:8.735e+08   3rd Qu.:2339329  
##  Max.   :65116403   Max.   :1.061e+09   Max.   :4282528  
##                                                          
##   UTR3.length       MultiTransGenes
##  Min.   :       0   Min.   : 3357  
##  1st Qu.: 1310786   1st Qu.: 3792  
##  Median : 3836127   Median : 4288  
##  Mean   : 6351649   Mean   : 5723  
##  3rd Qu.: 6916984   3rd Qu.: 7110  
##  Max.   :20893311   Max.   :11075  
##                     NA&#39;s   :1</code></pre>
<p>A data frame can be considered as a simple table. How do we add a new column? Lets make a column with average transcript lengths for all available genomes. We can then plot the results with a simple barplot function.</p>
<pre class="r"><code>genomes$Average.transcript.length &lt;- genomes$Transcripts.length/genomes$Transcripts

barplot(genomes$Average.transcript.length, names.arg=rownames(genomes), col=&quot;lightblue&quot;)</code></pre>
<p><img src="Intro_R_Practical_files/figure-html/unnamed-chunk-22-1.png" /><!-- --></p>
</div>
<div id="exercises" class="section level2">
<h2>Exercises</h2>
<ul>
<li>Make pie charts of exon versus intron length in each species (shown below)</li>
<li>Calculate length of an average human intron.</li>
<li>Which animal genomes have less than 20000 coding sequences?</li>
<li>How many exons per gene are on average for each of the available species?</li>
</ul>
<p>Lets make pie charts of exon vs intron total length for all genomes.</p>
<p>There are nine species, so we have to divide the plot into 9 parts.</p>
<pre class="r"><code>par(mfrow = c(3, 3))
for (species in rownames(genomes)) {
   print(species)
   pie(c(genomes[species,&quot;Coding.length&quot;],genomes[species,&quot;UTR5.length&quot;],genomes[species,&quot;UTR3.length&quot;]), main=species, labels=c(&quot;Coding&quot;,&quot;UTR5&quot;,&quot;UTR3&quot;))
}</code></pre>
<pre><code>## [1] &quot;cat&quot;</code></pre>
<pre><code>## [1] &quot;chicken&quot;</code></pre>
<pre><code>## [1] &quot;dog&quot;</code></pre>
<pre><code>## [1] &quot;drosophila&quot;</code></pre>
<pre><code>## [1] &quot;human&quot;</code></pre>
<pre><code>## [1] &quot;mouse&quot;</code></pre>
<pre><code>## [1] &quot;opossum&quot;</code></pre>
<pre><code>## [1] &quot;worm&quot;</code></pre>
<pre><code>## [1] &quot;zebrafish&quot;</code></pre>
<p><img src="Intro_R_Practical_files/figure-html/unnamed-chunk-23-1.png" /><!-- --></p>
<p>What percentage of genes has splice variants?</p>
<p>Use the function cor.test to test whether there is correlation between genome size and intron length. What happens when we eliminate the cat genome? (hint, use <strong>“genomes[-1,]”</strong> instead of <strong>“genomes”</strong>)</p>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
