---
title: "mRNA-Seq analysis"
author: "Anton Enright & Dimitrios Vitsios"
date: \'`r format(Sys.time(), '%d %B, %Y')`\'
always_allow_html: yes
output:
 github_document:
    toc: yes
---


```{r setup, results='hide',message=F, warning=F, echo=F}
require(knitr)
opts_knit$set(root.dir = '/Users/dvitsios/Desktop/EMBO_2017/Courses-and-Practicals/EMBO_Greece_2017/miR_210_Experiment/mRNA_Seq/')
opts_chunk$set(dev="png", 
               dev.args=list(type="cairo"),
               comment='           ##',
               dpi=300)

```


#mRNA Analysis 



## Raw Data
The raw data is [here](http://wwwdev.ebi.ac.uk/enright-srv/courses/rna_cambridge_2017/mrna_seq/data)

## Experimental Design

This experiment explores the effect of miR-210 Overexpression in MCF7 breast cancer cells (Human). 4 Lanes of mRNA-Seq data are used with 2 controls (overexpression of a C.elegans miR cel-mir-67) and 2 samples over-expressing miR-210.

This is the <em>pdata.txt</em> file:

|SampleName|fileName|condition|
|----------|---------|------------|
| m210\_1 | hisat.merged\_1\_16.fq.27197.bam.counts.merged|mir210|
| m210\_2 |hisat.merged\_1\_24.fq.27197.bam.counts.merged|mir210|
|cont\_1 | hisat.merged\_1\_32.fq.27197.bam.counts.merged|ctrl|
|cont\_2 |hisat.merged\_1\_40.fq.27197.bam.counts.merged|ctrl|


First we will load required modules and change to our working directory
```{r, message=F}
setwd('/Users/dvitsios/Desktop/EMBO_2017/Courses-and-Practicals/EMBO_Greece_2017/miR_210_Experiment/mRNA_Seq')
library(DESeq2)
library(gplots)
library(RColorBrewer)


```

Let’s make a colour palette (100 colours) for nice heatmaps (Green to Blue)


```{r}
hmcol = colorRampPalette(brewer.pal(9, 'GnBu'))(100)
```

## Reading in Annotation and Count Data

We’ll now read in the experimental design file (pdata.txt) and the a containing gene names (gene_names.txt).
We will also create a vector containing condition names for each sample.
Finally, we will make a vector of colours that correspond to each condition for nicer plots later on. We’ll also make colours corresponding to each class (<em>biotype</em>) of gene.

```{r, results="hold"}
pdata=read.table('data/pdata.txt',header=TRUE)

conds=as.vector(pdata$condition) 
names=read.table('data/gene_names.txt',header=FALSE,row.names=1)

condcols=brewer.pal(n = length(unique(conds))+2, name = 'Dark2')
names(condcols)=unique(conds)

typecols=rainbow(44)[names$V4]
names(typecols)=rownames(names)
```

Next we’ll read in the actual raw counts derived from HTSeq.
The DESeq2 function has an option <em>DESeqDataSetFromHTSeqCount</em> that reads in counts from HTSeq output files. We just need to provide it with a directory and an appopropriate <em>pdata.txt</em> file.


```{r, results="hold"}
ddsHTSeq = DESeqDataSetFromHTSeqCount(sampleTable = pdata, directory = '.', design= ~ condition)

# Restrict to protein coding and lincRNA ?
#ddsHTSeq=ddsHTSeq[(names[rownames(ddsHTSeq),'V4'] == protein_coding) | (names[rownames(ddsHTSeq),'V4'] == lincRNA),]

colData(ddsHTSeq)$condition = factor(colData(ddsHTSeq)$condition,levels=levels(pdata$condition))
```

## Data Normalisation

```{r, results="hold"}
# Data Normalisation
dds = estimateSizeFactors(ddsHTSeq)
# Data Dispersion
dds = estimateDispersions(dds)
# New Neg. Binomial test
dds = nbinomWaldTest(dds)

counts_table=counts(dds,normalized=TRUE)
```

Let’s look at the DESeq2 Dispersion plot to see how it looks.
```{r}
plotDispEsts(dds,main='Dispersion Plot')
```


## Post-Normalisation QC

```{r, results="hold"}
par(mfrow=c(3,1))
barplot(colSums(counts(dds, normalized=F)), col=condcols[as.factor(conds)], las=2,cex.names=0.4,main='Pre Normalised Counts')
plot(1, type='n', axes=F, xlab='', ylab='')
legend('center',levels(as.factor(conds)),fill=condcols)
barplot(colSums(counts(dds, normalized=T)), col=condcols[as.factor(conds)], las=2,cex.names=0.4,main='Post Normalised Counts')
```

### Stabilising the Variance (VST) instead of log2

Regularised log transformation and variance stabilising transformation, these function are usesful to visualise the data. See the DESeq2 manual for more information.

```{r, results="hold"}
vsd = varianceStabilizingTransformation(dds)
vstMat = assay(vsd)
counts_table=data.frame(counts_table,vstMat)
```


### Sample to Sample Correlation
```{r}
heatmap.2(cor(assay(vsd)),trace='none',main='Sample Correlation Variance Stabilised',col=hmcol,cexRow=0.6,cexCol=0.6)
```


### Sample to Sample PCA
```{r, results="hold"}
# Sample PCA

pca = princomp(assay(vsd))
plot(pca$loadings, main='PCA Variance Stabilised', col='black', bg=condcols[pdata$condition],  pch=21, cex=1)
text(pca$loadings, conds, pos=3, cex=0.8)
```


## Statistical Analysis

Now we will perform the statistical tests.
1. Control vs miR-210 Overexpression

For the test we are looking for how many genes are significantly changed with an Absolute Log Fold Change of at least 1 and an Adjusted P-Value of less than 0.05.
Lets see how many genes we find.
```{r, results="hold"}
res = results(dds, contrast=c('condition', 'ctrl', 'mir210'))
res=res[order(res$padj),]

res
```

```{r}
hits=(rownames(res[((res$padj <= 0.05) & (abs(res$log2FoldChange) >= 1) & (!is.na(res$padj))),]))
```

## Analysis of Sample Median Data

Sometimes it is useful to compute the median of each sample so we can make some different plots (e.g. pairwise plots).
```{r, results="hold"}
ctr_median=apply(vstMat[,3:4],1,median)
m210_median=apply(vstMat[,1:2],1,median)

plot(ctr_median,m210_median,pch=19,main=paste('ctrl vs m210'),col='darkblue',cex=0.4)
points(ctr_median[hits],m210_median[hits],pch=19,col=typecols[hits],cex=0.4)
text(ctr_median[hits],m210_median[hits],cex=0.4,pos=4,labels=names[hits,'V2'])
abline(a=0,b=1,lty=2,col='red',lwd=2,cex=0.4)
```

```{r, results="hold"}
plot(ctr_median,m210_median,pch=19,main=paste('ctrl vs m210 Biotype'),col='darkblue',cex=0.4)
points(ctr_median[hits],m210_median[hits],pch=19,col=typecols[hits],cex=0.4)
points(ctr_median,m210_median,pch=19,col=typecols,cex=0.2)
text(ctr_median[hits],m210_median[hits],cex=0.4,pos=4,labels=names[hits,'V2'])
abline(a=0,b=1,lty=2,col='red',lwd=2,cex=0.4)
legend('topleft',levels(unique(names$V4)),cex=0.4,bty='n',fill=rainbow(44))
```


## Post-Statistical QC

Now we will construct a Volcano Plot to highlight each gene’s Significance versus its Fold Change. Genes which do not change much usually have poor P-Values so the plot has low significance around zero and hence looks like a <em>volcano</em>. Some genes may have high fold changes but poor significance, in these cases, usually the foldchanges observed do not replicate well from sample to sample.

### Volcano Plot of Significance and Fold Change

```{r, results="hold"}
plot(res$log2FoldChange,-log(res$padj,10),ylab='-log10(Adjusted P)',xlab="Log2 FoldChange", main=paste('Volcano Plot Ctrl v miR210_oe'),pch=19,cex=0.2)      
text(res[hits,]$log2FoldChange,-log(res[hits,]$padj,10),labels=names[rownames(res[hits,]),'V2'],pos=3,cex=0.4)
points(res[hits,'log2FoldChange'],-log(res[hits,'padj'],10),pch=19,cex=0.4,col=typecols[hits])
legend('topleft',"Up in 210\nDown in Ctrl",cex=0.5)
legend('topright',"Down in 210\nUp in Ctrl",cex=0.5)
abline(h=-log10(0.05),lty=3)
abline(v=-1,lty=3)
abline(v=1,lty=3)
```

### Heatmap of Signficant Hits

We now compute a heatmap of the most significant hits.

```{r}
heatmap.2(vstMat[hits,],trace='none',col=hmcol,labRow=names[hits,'V2'],cexRow=0.4,cexCol=0.6,las=2,Colv=FALSE,dendrogram='row',main='Heatmap Sig. Hits', RowSideColors = typecols[hits])
```


## Output Final Results

Let’s write the results out to a txt table file called <strong>FullCountsandStats.txt</strong>.

```{r, results="hold"}
fullresults=merge(names,counts_table,by.x=0,by.y=0)
fullresults=merge(fullresults,as.matrix(res),by.x=1,by.y=0)
fullresults=fullresults[order(fullresults$log2FoldChange,decreasing=T),]
fullresults=fullresults[!is.na(fullresults$log2FoldChange),]
rownames(fullresults)=fullresults$Row.names

write.table(fullresults,"FullCountsandStats.txt",sep="\t",quote=F)
```



