Wellcome Trust Advanced Courses - RNA Transcriptomics
===============================
![WTAC](/images/acsc_logo.png)

![Wellcome](/images/wellcome_logo.png)
![Cambridge](/images/cambridge.jpg)

2018 WTAC - RNA Transcriptomics, Hinxton, Cambridge, UK
-------------------------------------------------------------------

**Instructors: Anton Enright, Jack Monahan, Adrien Leger**

|Anton Enright (aje39@cam.ac.uk) | Jack Monahan (monahanj@ebi.ac.uk) | Adrien Leger (aleg@ebi.ac.uk)|
|---------------------------|------------------------------------|------------------------------------|
|<img src="../images/anton.jpg" height="150">|<img src="../images/jack_picture.png" height="150">|<img src="../images/adrien.jpeg" height="150">|

***

<img src="../images/cambridge.jpg" align="right" width="150">

_Department of Pathology,  
Tennis Court Road,  
Cambridge, UK._  

***

List of Practicals:
------------------

1. [Introduction to R/BioConductor](Intro_R/Intro_R_Practical.md)

2. Basics of SmallRNA Sequencing Data Analysis
   * [SmallRNA Sequencing - Basic Analysis and QC](small_RNA_seq/Practical_1/Practical_1.md)
   * [SmallRNA Sequencing - Mapping and Statistical analysis of count data](small_RNA_seq/Practical_2/Practical_2.md)

3. Worked Example for microRNA Target Analysis - MiR-210 in Breast Cancer
   * [SmallRNA Sequencing - Breast Cancer miR210 Samples - Participant Driven](miR_210_Experiment/small_RNASeq/small_RNASeq.md)
   * [Mini mapping practical using HISAT2](miR_210_Experiment/mini_mapping/)
   * [mRNA Sequencing - Breast Cancer miR210 Samples](miR_210_Experiment/mRNA_Seq/mRNA_Seq.md)
   
4. Worked Example for microRNA Target Analysis - MiR-210 in Breast Cancer
   * [SmallRNA Sequencing - Course Data 2018 - Clean, Map, QC, Analysis](course_smallrna_seq/course_smallrna.md)
   
5.  [MicroRNA Target Analysis using mRNASeq and smallRNA Seq combined](Sylamer_microRNA_targets)

6.  [_de novo_ prediction of microRNAs from sequencing data](mirnovo)

7. [Nanopore Direct RNA sequencing  analysis](Nanopore_dRNA_Seq)

***

Useful resources:

* [Enright Lab](http://www.ebi.ac.uk/research/enright)
* [R Cran](https://cran.r-project.org/)
* [R Studio](http://www.rstudio.com/)
* [BioConductor](http://www.bioconductor.org)
* [DESeq2 Manual](http://bioconductor.org/packages/release/bioc/html/DESeq2.html)
* [ChimiRa Webserver](http://wwwdev.ebi.ac.uk/enright-dev/chimira/)
* [miRnovo Webserver](http://wwwdev.ebi.ac.uk/enright-dev/mirnovo/)
* [miRBase](http://www.mirbase.org)
* [Legacy Course Material, pre 2017](http://wwwdev.ebi.ac.uk/enright-srv/courses)
