<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Anton Enright &amp; Dimitrios Vitsios" />


<title>Small RNA Seq - Practical 2</title>

<script src="Practical_2_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="Practical_2_files/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="Practical_2_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="Practical_2_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="Practical_2_files/bootstrap-3.3.5/shim/respond.min.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<link rel="stylesheet"
      href="Practical_2_files/highlight/textmate.css"
      type="text/css" />
<script src="Practical_2_files/highlight/highlight.js"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>



<div class="container-fluid main-container">

<!-- tabsets -->
<script src="Practical_2_files/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->






<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Small RNA Seq - Practical 2</h1>
<h4 class="author"><em>Anton Enright &amp; Dimitrios Vitsios</em></h4>
<h4 class="date"><em>'12 June, 2017'</em></h4>

</div>

<div id="TOC">
<ul>
<li><a href="#differential-expression-of-smallrna-counts-with-deseq2">Differential Expression of smallRNA counts with DESeq2</a></li>
<li><a href="#initial-count-analysis">Initial count analysis</a><ul>
<li><a href="#using-deseq-to-normalise-the-smallrna-count-data">Using DESeq to normalise the smallRNA count data</a></li>
<li><a href="#post-normalisation-analysis">Post Normalisation Analysis</a></li>
<li><a href="#analysis-of-significant-hits-across-all-samples">Analysis of Significant Hits across all samples</a></li>
</ul></li>
</ul>
</div>

<div id="differential-expression-of-smallrna-counts-with-deseq2" class="section level1">
<h1>Differential Expression of smallRNA counts with DESeq2</h1>
<p>First we should change directory to where the data is</p>
<pre class="r"><code>setwd(&quot;/Users/aje/Desktop/Courses-and-Practicals/EMBO_Greece_2017/small_RNA_seq/data&quot;)</code></pre>
<p>If you need the raw data for this practical, it is available <a href="http://wwwdev.ebi.ac.uk/enright-srv/courses/rna_cambridge_2017/miseqdeseq/data">here</a></p>
<pre class="r"><code>library(&quot;DESeq2&quot;)
library(&quot;gplots&quot;)
library(&quot;RColorBrewer&quot;)

# Make a color scheme for heatmaps
hmcol = colorRampPalette(brewer.pal(9, &quot;GnBu&quot;))(100)</code></pre>
<p>We will load in the counts obtained from BLAST of cleaned reads against miRBase mature human sequences</p>
<pre class="r"><code>mircounts &lt;- read.table(&quot;mircounts.txt&quot;,header=TRUE,row.names=1)

# We want to remove the last line of the mircounts file
mircounts &lt;- mircounts[-nrow(mircounts),]

# We need to tell R which samples were from which tissues
samplenames &lt;- c(&quot;Brain&quot;, &quot;Brain&quot;, &quot;Brain&quot;, &quot;Heart&quot;, &quot;Heart&quot;, &quot;Heart&quot;, &quot;Heart&quot;, &quot;Heart&quot;, &quot;Liver&quot;, &quot;Liver&quot;, &quot;Liver&quot;, &quot;Liver&quot;)
colnames(mircounts) = paste(samplenames, seq(1:12), sep=&quot;_&quot;)
head(mircounts)</code></pre>
<pre><code>##                 Brain_1 Brain_2 Brain_3 Heart_4 Heart_5 Heart_6 Heart_7
## hsa-miR-1          1167    1911    1190  179717  129698  114090  164659
## hsa-miR-122-5p      145     349     156     295      95     187     155
## hsa-miR-9-5p      86907   98965   74695     444     124     296     239
## hsa-miR-143-3p    27598   23711   21600   78973   66404   68549   96678
## hsa-miR-148a-3p    2079    2053    1646    6637    4126    4034    6276
## hsa-miR-21-5p      8279    8322    7029   14414    9933   11397   14733
##                 Heart_8 Liver_9 Liver_10 Liver_11 Liver_12
## hsa-miR-1          1253     375      554      345      348
## hsa-miR-122-5p      183  157280    73426   145594    63610
## hsa-miR-9-5p      84410     136      256      134      137
## hsa-miR-143-3p    24126   16411     8800    10078     7491
## hsa-miR-148a-3p    1697   73860    40933    52708    32330
## hsa-miR-21-5p      6732   58983    28787    32407    26526</code></pre>
</div>
<div id="initial-count-analysis" class="section level1">
<h1>Initial count analysis</h1>
<p>First, lets see the total numbers of counts obtained for each sample. We will use the <strong>apply</strong> function to quickly sweep across the table and compute the column sums.</p>
<pre class="r"><code>barplot(apply(mircounts,2,sum),col=as.factor(samplenames),las=2)</code></pre>
<p><img src="Practical_2_files/figure-html/unnamed-chunk-4-1.png" /><!-- --></p>
<p>Some of the samples look dramatically different to their replicates. We should investigate further by comparing samples to each other.</p>
<p>First we’ll do a pairwise plot of the log2 counts between all samples</p>
<pre class="r"><code>pairs(log2(mircounts+1),main=&quot;Pair-wise sample to sample counts&quot;)</code></pre>
<p><img src="Practical_2_files/figure-html/unnamed-chunk-5-1.png" /><!-- --></p>
<p>Does anything look fishy about the data to you ?. Let’s look at how the samples correlate with each other. Obviously replicates should be very highly correlated with a standard Pearson correlation test.</p>
<pre class="r"><code>pca &lt;- princomp(mircounts)
plot(pca$loadings, col=as.factor(samplenames),  pch=19, cex=2, main=&quot;Sample to Sample&quot;)
text(pca$loadings, as.vector(colnames(mircounts)), pos=3, cex=0.8)</code></pre>
<p><img src="Practical_2_files/figure-html/unnamed-chunk-6-1.png" /><!-- --></p>
<pre class="r"><code>heatmap.2(cor(mircounts),trace=&quot;none&quot;,col=hmcol,main=&quot;Sample Correlation&quot;)</code></pre>
<p><img src="Practical_2_files/figure-html/unnamed-chunk-7-1.png" /><!-- --></p>
<p>Due to the sample-swap error we need to relabel the swapped samples</p>
<pre class="r"><code>samplenames &lt;- c(&quot;Brain&quot;, &quot;Brain&quot;, &quot;Brain&quot;, &quot;Heart&quot;, &quot;Heart&quot;, &quot;Heart&quot;, &quot;Heart&quot;, &quot;Brain&quot;, &quot;Liver&quot;, &quot;Liver&quot;, &quot;Liver&quot; ,&quot;Liver&quot;)</code></pre>
<p>Let’s double check things are now fixed</p>
<pre class="r"><code>pca &lt;- princomp(mircounts)
plot(pca$loadings, col=as.factor(samplenames),  pch=19, cex=2, main=&quot;Sample to Sample&quot;)
text(pca$loadings, as.vector(colnames(mircounts)), pos=3, cex=0.8)</code></pre>
<p><img src="Practical_2_files/figure-html/unnamed-chunk-9-1.png" /><!-- --></p>
<p>Clearly we need to normalise the data to control for differences in global RNA levels across samples.</p>
<div id="using-deseq-to-normalise-the-smallrna-count-data" class="section level2">
<h2>Using DESeq to normalise the smallRNA count data</h2>
<p>DESeq is a statistical tool for analysis of count-based data such as from RNAseq. Microarrays and similar platforms produce ‘continuous’ data measurements, e.g. flourescence associated with a probe. However for count data the variance of results increases dramatically as you get low counts. For example, once a gene is lowly enough expressed that you only find small numbers of reads mapping to it you get very high variance as it is impossible to have half a count. For this reason it is imperative that count based sequencing data be normalised and statistically assessed with tools that take this into account. Tools like DESeq apply negative binomial statistics and try to flatten the variance across low and high counts.</p>
<pre class="r"><code># First we tell DESeq which samples correspond to which tissues.
conds = data.frame(samplenames)
colnames(conds)=&quot;tissue&quot;

# Now we build a DESeq Count dataset and normalize it.
cds &lt;- DESeqDataSetFromMatrix(countData = mircounts, colData = conds, design = ~ tissue)
cds &lt;- estimateSizeFactors(cds)
cds &lt;- estimateDispersions(cds)</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre class="r"><code>cds &lt;- nbinomWaldTest(cds)</code></pre>
<p>Now we will plot the dispersion information and fit.</p>
<pre class="r"><code>plotDispEsts(cds)</code></pre>
<p><img src="Practical_2_files/figure-html/unnamed-chunk-11-1.png" /><!-- --></p>
</div>
<div id="post-normalisation-analysis" class="section level2">
<h2>Post Normalisation Analysis</h2>
<p>Lets see what effect our normalisation had</p>
<pre class="r"><code>par(mfrow=c(2,1))
prenorm=apply(mircounts,2,sum)
barplot(prenorm,col=as.factor(samplenames),las=2,names=samplenames)
postnorm=apply(counts(cds,normalized=TRUE),2,sum)
barplot(postnorm,col=as.factor(samplenames),las=2,names=samplenames)</code></pre>
<p><img src="Practical_2_files/figure-html/unnamed-chunk-12-1.png" /><!-- --></p>
<p>Lets do another Principal components analysis on the normalised data</p>
<pre class="r"><code>pca &lt;- princomp(counts(cds,normalized=T))
plot(pca$loadings, col=as.factor(samplenames),  pch=19, cex=2, main=&quot;Sample to Sample PCA&quot;)
text(pca$loadings, as.vector(colnames(mircounts)), pos=3, cex=0.8)</code></pre>
<p><img src="Practical_2_files/figure-html/unnamed-chunk-13-1.png" /><!-- --></p>
<p>Now we can use the negative-binomial test for each pairwise comparison of interest.</p>
<pre class="r"><code>res1 =  results( cds, contrast=c(&quot;tissue&quot;,&quot;Brain&quot;, &quot;Heart&quot;))
res2 =  results( cds, contrast=c(&quot;tissue&quot;,&quot;Brain&quot;, &quot;Liver&quot;))
res3 =  results( cds, contrast=c(&quot;tissue&quot;,&quot;Heart&quot;, &quot;Liver&quot;))

# Sort each result on Adjusted P-Value
res1&lt;-res1[order(res1$padj),]
res2&lt;-res2[order(res2$padj),]
res3&lt;-res3[order(res3$padj),]

# Look at the first comparison
head(res1,50)</code></pre>
<pre><code>## log2 fold change (MAP): tissue Brain vs Heart 
## Wald test p-value: tissue Brain vs Heart 
## DataFrame with 50 rows and 6 columns
##                   baseMean log2FoldChange      lfcSE      stat
##                  &lt;numeric&gt;      &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;
## hsa-miR-128      9917.6289       4.716466 0.11809294  39.93859
## hsa-miR-378a-3p 12118.3791      -6.247371 0.09553713 -65.39207
## hsa-miR-499a-5p  7258.3710      -7.021495 0.07591700 -92.48910
## hsa-miR-133a     4807.4627      -7.940650 0.16459842 -48.24256
## hsa-miR-378d      928.9525      -6.269205 0.15571567 -40.26059
## ...                    ...            ...        ...       ...
## hsa-miR-136-3p   162.83485       3.245716  0.1623744  19.98908
## hsa-miR-452-5p    75.61569      -4.610875  0.2325850 -19.82447
## hsa-miR-9-3p     227.13253       6.624428  0.3393934  19.51844
## hsa-miR-30b-5p   711.24239      -2.521038  0.1292027 -19.51226
## hsa-miR-744-5p   252.02968       2.541499  0.1316901  19.29909
##                       pvalue         padj
##                    &lt;numeric&gt;    &lt;numeric&gt;
## hsa-miR-128                0            0
## hsa-miR-378a-3p            0            0
## hsa-miR-499a-5p            0            0
## hsa-miR-133a               0            0
## hsa-miR-378d               0            0
## ...                      ...          ...
## hsa-miR-136-3p  6.854389e-89 1.068391e-87
## hsa-miR-452-5p  1.830961e-87 2.793189e-86
## hsa-miR-9-3p    7.654455e-85 1.143384e-83
## hsa-miR-30b-5p  8.637456e-85 1.263889e-83
## hsa-miR-744-5p  5.466404e-83 7.838824e-82</code></pre>
<pre><code>log2 fold change (MAP): tissue Brain vs Heart 
Wald test p-value: tissue Brain vs Heart 
DataFrame with 50 rows and 6 columns
                  baseMean log2FoldChange      lfcSE      stat       pvalue         padj
                 &lt;numeric&gt;      &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt;    &lt;numeric&gt;hsa-miR-128      9917.6289       4.716332 0.11968367  39.40664            0            0
hsa-miR-378a-3p 12118.3791      -6.247263 0.09765315 -63.97401            0            0
hsa-miR-499a-5p  7258.3710      -7.021408 0.07956721 -88.24500            0            0
hsa-miR-133a     4807.4627      -7.940515 0.16499010 -48.12722            0            0
hsa-miR-378d      928.9525      -6.269041 0.15687156 -39.96289            0            0
...                    ...            ...        ...       ...          ...          ...
hsa-miR-136-3p    162.8349       3.245689 0.16159117  20.08581 9.821032e-90 1.603390e-88
hsa-miR-9-3p      227.1325       6.624195 0.33683195  19.66617 4.203540e-86 6.716720e-85
hsa-miR-27b-3p  20353.5151      -1.448201 0.07405947 -19.55457 3.771937e-85 5.901510e-84
hsa-miR-30b-5p    711.2424      -2.521019 0.12998167 -19.39519 8.474115e-84 1.298788e-82
hsa-miR-744-5p    252.0297       2.541474 0.13180332  19.28232 7.560116e-83 1.135529e-81</code></pre>
<p>Lets make some volcanoplots of each comparison</p>
<pre class="r"><code>par(mfrow=c(1,3))
plot(res1$log2FoldChange,-log(res1$padj,10),main=&quot;Volcano Plot Brain vs Heart&quot;)
text(res1[1:20,]$log2FoldChange,-log(res1[1:20,]$padj,10),labels=rownames(res1[1:20,]),cex=0.7,pos=1)
legend(&quot;topleft&quot;,&quot;Brain&quot;,cex=0.5)
legend(&quot;topright&quot;,&quot;Heart&quot;,cex=0.5)

plot(res2$log2FoldChange,-log(res2$padj,10),main=&quot;Volcano Plot Brain vs Liver&quot;)
text(res2[1:20,]$log2FoldChange,-log(res2[1:20,]$padj,10),labels=rownames(res2[1:20,]),cex=0.7,pos=1)
legend(&quot;topleft&quot;,&quot;Brain&quot;,cex=0.5)
legend(&quot;topright&quot;,&quot;Liver&quot;,cex=0.5)

plot(res3$log2FoldChange,-log(res3$padj,10),main=&quot;Volcano Plot Heart vs Liver&quot;)
text(res3[1:20,]$log2FoldChange,-log(res3[1:20,]$padj,10),labels=rownames(res3[1:20,]),cex=0.7,pos=1)
legend(&quot;topleft&quot;,&quot;Heart&quot;,cex=0.5)
legend(&quot;topright&quot;,&quot;Liver&quot;,cex=0.5)</code></pre>
<p><img src="Practical_2_files/figure-html/unnamed-chunk-15-1.png" /><!-- --></p>
<pre class="r"><code>par(mfrow=c(1,1))</code></pre>
</div>
<div id="analysis-of-significant-hits-across-all-samples" class="section level2">
<h2>Analysis of Significant Hits across all samples</h2>
<p>Let’s choose significant miRs for each contrast by log fold change and adj. P-value. Then we merge into a single list of significant hits and make a heatmap.</p>
<pre class="r"><code>sig1 = rownames(res1[(abs(res1$log2FoldChange) &gt; 4) &amp; (res1$padj &lt; 0.00001) &amp; !is.na(res1$padj),])
sig2 = rownames(res2[(abs(res2$log2FoldChange) &gt; 4) &amp; (res2$padj &lt; 0.00001) &amp; !is.na(res2$padj),])
sig3 = rownames(res3[(abs(res3$log2FoldChange) &gt; 4) &amp; (res3$padj &lt; 0.00001) &amp; !is.na(res3$padj),])

# Merge to one list
siglist = unique(c(sig1,sig2,sig3))

# Generate Nice heatmap colours
hmcol = colorRampPalette(brewer.pal(9, &quot;GnBu&quot;))(100)

# Heatmap of significant hits
heatmap.2(log2(counts(cds[siglist,],normalized=TRUE)+1),col=hmcol,trace=&quot;none&quot;,labCol=samplenames,margin=c(5,10))</code></pre>
<p><img src="Practical_2_files/figure-html/unnamed-chunk-16-1.png" /><!-- --></p>
<p>We can also make a more simplified heatmap of expression for 20 most significant hits from each comparison.</p>
<pre class="r"><code>siglist=unique(c(rownames(res1[1:20,]),rownames(res2[1:20,]),rownames(res3[1:20,])))

heatmap.2(log2(counts(cds[siglist,],normalized=TRUE)+1),col=hmcol,trace=&quot;none&quot;,margin=c(5,10))</code></pre>
<p><img src="Practical_2_files/figure-html/unnamed-chunk-17-1.png" /><!-- --></p>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
